### DISCUSSION TITLE
demo/with-cra 에서 yarn 빌드 오류

### DISCUSSION THREAD

**Discussion by hy57in** [EveryAnalytics/react-analytics-provider#204] [General]
## Issue description

demo/with-cra 폴더 내에서 개발서버를 띄우거나 빌드시 오류가 납니다ㅜ
demo/with-cra 폴더 내에서 yarn install 해도 중간에 오류가 뜹니다ㅜ

yarn.lock 지우고 다시 해봐도 안되네용

config에 오류가 있는듯한데 건들 수가 없어 도움을 요청ㅇ합니다!

## Steps to reproduce the issue

1. root에서 `yarn install && yarn serve`
2. /demo/with-cra 에 들어간다
3. `yarn install`
4. `yarn dev`

## What's the expected result?

- yarn dev 시 개발서버 동작..

## What's the actual result?

- yarn install 하면
```shell
➤ YN0001: │ Error: @types/react-dom@file:../../node_modules/react-dom::locator=with-nextjs%40workspace%3A.: ENOENT: no such file or directory, lstat
```
- yarn dev 또는 yarn build 하면
```shell
Internal Error: demo@workspace:.: This package doesn't seem to be present in your lockfile; run "yarn install" to update the lockfile
```
![image](https://user-images.githubusercontent.com/60775453/131868999-f4f4c453-7295-4db6-a923-777840ae9cb8.png)

## Additional details

-

**Comment by greatSumini** [comment]
올려주신 에러는 RAP 패키지 관리자가 Yarn berry로 변경되면서 루트 경로의 node_modules가 사라져서 발생한 것 같습니다 ㅎㅎ..
근데 react, react-dom을 따로 설치해서 해당 에러를 해결해봐도 serve중인 RAP의 dependency들 못 읽네요 ㅠㅠ `node_modules`의 느슨한 의존성 관리 덕분에 정상 작동하고 있던 것 같습니다 😭 

저희가 해볼 수 있는 조치들은 아래와 같이 있는듯한데.. 논의가 필요할 것 같아용

1. RAP에서 Yarn berry를 제거하고 다시 `node_modules`를 사용한다.
2. demo의 디펜던시들까지 아예 하나의 패키지로 관리한다.
3. [vercel/next.js examples](https://github.com/vercel/next.js/blob/canary/examples/active-class-name/package.json)처럼 demo 프로젝트들의 relative dependency들을 모두 제거해 RAP와 완전히 격리한다. (지금처럼 개발 모드 켜놓고 모니터링하는 용도로 사용할 수 없게 됨)
4. yarn workspace를 도입한다. (workspace 도입하려면 RAP도 통째로 옮기는게 맞을듯한디.. 이건 좀 아닌둣합니다)

개인적으로는 1번을 선택해 빠르게 문제 해결하고 신규 기능 추가에 집중하는 것도 좋을듯합니다 ㅎㅎ..

**Comment by hy57in** [comment]
> 올려주신 에러는 RAP 패키지 관리자가 Yarn berry로 변경되면서 루트 경로의 node_modules가 사라져서 발생한 것 같습니다 ㅎㅎ..
> 근데 react, react-dom을 따로 설치해서 해당 에러를 해결해봐도 serve중인 RAP의 dependency들 못 읽네요 ㅠㅠ `node_modules`의 느슨한 의존성 관리 덕분에 정상 작동하고 있던 것 같습니다 😭
> 
> 저희가 해볼 수 있는 조치들은 아래와 같이 있는듯한데.. 논의가 필요할 것 같아용
> 
> 1. RAP에서 Yarn berry를 제거하고 다시 `node_modules`를 사용한다.
> 2. demo의 디펜던시들까지 아예 하나의 패키지로 관리한다.
> 3. [vercel/next.js examples](https://github.com/vercel/next.js/blob/canary/examples/active-class-name/package.json)처럼 demo 프로젝트들의 relative dependency들을 모두 제거해 RAP와 완전히 격리한다. (지금처럼 개발 모드 켜놓고 모니터링하는 용도로 사용할 수 없게 됨)
> 4. yarn workspace를 도입한다. (workspace 도입하려면 RAP도 통째로 옮기는게 맞을듯한디.. 이건 좀 아닌둣합니다)
> 
> 개인적으로는 1번을 선택해 빠르게 문제 해결하고 신규 기능 추가에 집중하는 것도 좋을듯합니다 ㅎㅎ..

오호 yarn berry 제거하는게 쉽다면 되돌려도 좋을거같아요! 그런데 나중에 다시 yarn berry를 도입했을 때 결국 2,3,4 중 하나의 방법으로 문제를 해결해야하는거죠?

**Comment by jhlee910609** [comment]
@greatSumini 이상적인 건 4번인데 쉽게 될지는 미지수네요 ㅠㅠㅠ

**Comment by greatSumini** [comment]
@hy57in 네 그춍 근데 Yarn Berry가 꼭 필요하진 않을 수도 있을 것 같아여 배포가 잦지 않고 설치된 의존성이 적어 굳이..? 일수도 있을 것 같습니다!

**Comment by greatSumini** [comment]
@jhlee910609 저는 개발서버 테스트는 스토리북이나 E2E테스트로 해결하고 3번처럼 데모는 완전격리하는게 좋다고보긴하는디 환경 구성이 조금 어렵긴하져 ㅎ.ㅠ

**Comment by jhlee910609** [comment]
구글링 해도 [yarn link](https://classic.yarnpkg.com/en/docs/cli/link/) 직접 해주 방법 외에는 잘 모르겠어서 몇 일 전에 yarn berry discussion에 문의했습니다. pnp mode에서 module linking 하는 방식이 있을까 해서요. 저는 해당 discussion 트래킹하고 있을게요.
https://github.com/yarnpkg/berry/discussions/3379

**Comment by josworks27** [comment]
제 생각도 demo는 rap에서 완전 분리되는 편이 좋지 앟을까 싶습니다.
현재는 file: .... 로 의존성을 찾는 방법을 사용하고 있는데, 이는 demo와 rap의 개발이 병렬로 진행되고 있어서 발생한 이슈 같습니다.

먼저 rap의 1.0.0 개발이 완료된 후 npm 배포 후 demo를 구현하는 방식은 어떨까요 ?? 이렇게 했을 시 아래와 같은 이점이 있을 것 같습니다.
1. 참여하는 인원들의 리소스를 rap에 집중시킬 수 있다.
2. rap가 변경됨에 따라 demo도 같이 바꿔야하는데 이러한 유지보수 관점에서의 관리 포인트를 줄일 수 있다.

죄송하게도 제가 demo는 많은 관심을 가지지 못하여 현재 이슈에 대한 이해가 부족합니다. 많이 의견 말씀해주시면 배우는 자세로 참고하겠습니다 !

**Comment by milooy** [comment]
@josworks27 님이 말씀해주신

> 2. rap가 변경됨에 따라 demo도 같이 바꿔야하는데 이러한 유지보수 관점에서의 관리 포인트를 줄일 수 있다.

관점에서 demo앱에서 local dependency를 떼는게 좋겠네요.
현시점에서 0.2.0으로 릴리즈하고 demo 분리시켜볼까요?

**Reply by jhlee910609** [reply]
@milooy @chanheeis RAP/demo directory 별도로 분리하는게 여러 측면에서 좋아 보입니다. 근본적인 해결책 같기도 하구요. 한 RAP directory의 sub directory로 관리할 필요성(RAP 하위 directory에서 관리할 이유)도 좀 적어 보입니다!

**Reply by milooy** [reply]
레포는 하나로 둬도 좋지 않을까요? 어떤 측면에서 분리가 좋을거같으셔요?

**Reply by josworks27** [reply]
레포 분리의 필요성이 있다면 monorepo 방식으로 하는 것도 고려해볼 수 있는 선택지일 것 같습니다 !

**Reply by greatSumini** [reply]
저는 [vercel/next.js - examples](https://github.com/vercel/next.js/tree/canary/examples)처럼 같은 repo의 sub directory로 관리하되 모듈 의존성은 완전히 격리하는 것도 좋을 것 같습니다! 이유는 다음과 같습니다.
1. RAP라는 맥락을 공유하기 때문에 한 repo에서 관리되는 것이 자연스럽다. (찾기 쉽고, 관리하기 쉽다.)
2. 공유되는 의존성이 적고 그 연관관계가 모호하기 때문에 Lerna, Yarn workspace 등의 모노레포 도구 도입은 조금 부적절하다.

**결론**: 패키지 관리는 따로 하고, 레포지토리만 공유하는 사전적 의미?의 모노레포 정도만 유지하면 어떨까싶습니다 ㅎㅎ

**Reply by jhlee910609** [reply]
@milooy 아, 제가 표현을 잘못한 것 같습니다.ㅠㅠ
@greatSumini 님께서 말씀하신 것 처럼 sub directory 형식을 말씀드린거 였는데, 저 글 작성할 때 졸렸나 봅니다... 😢  
![image](https://user-images.githubusercontent.com/28521282/132144236-fc5ca4e0-00a2-4f19-bf50-c1152b44146f.png)

**Reply by jhlee910609** [reply]
추가적으로 수민님 말씀처럼 RAP와 demo app간의 공유할 의존성이 적으니 monorepo나 yarn workspace 적용이 의미가 클지 잘 모르겠습니다. demo 안에서는 /with-cra, /with-nextjs 두 repo 간에 어느 정도 의존성 공유하는 것들이 있으니 yarn workspace정도는 시도해볼 수 있을 것 같네요! 이 부분은 제가 시도해보고 싶기도 합니다~

**Reply by hiphapis** [reply]
@jhlee910609 님 시도해보시고, 말씀주시는건 어떨까요?. 많은 시간을 쓰시진 말고 살짝쿵 해보고 안된다 싶으면 다시 말씀주시는걸로

**Reply by jhlee910609** [reply]
@hiphapis 네, 일단 0.2.0 버전 배포 후(?) 위 내용 보고 한번 시도해보고 어려우면 다시 끌올해보겠습니다~ 
@greatSumini 도와주시면 감사하겠습니다 ㅠ

**Reply by milooy** [reply]
굿굿 
대략 아래처럼 의견 모아진것같은데 확인 부탁드려요~

1. 0.2.0 릴리즈 @milooy  (DONE in https://github.com/EveryAnalytics/react-analytics-provider/pull/207)
2. demo에서 local dependency 떼기 
3. demo에 yarn workspaces 사용해보기 @jhlee910609 

2번 이슈 관심 있으신분? 여기도 @jhlee910609 님께서?

**Reply by milooy** [reply]
아근데 demo에서 local dependency가 떼지면
라이브러리 개발이 불편해지겠네요
test case 작성 외에 어떻게 live하게 개발할 수 있을까요 🤔

**Reply by jhlee910609** [reply]
제가 정확히는 모르지만 2번이 지금 제일 큰 일인 것 같은데요.ㅠㅠㅠ 3번은 급한 이슈는 아니라 우선순위는 낮아 보여서요. 흠.... yarn berry 사용 시, module path를 linking 해줄 수 있는지 더 조사해볼게요.

**Reply by milooy** [reply]
앗? 2번 그냥 

as-is
```
"dependencies": {
    "@every-analytics/react-analytics-provider": "file:../..",
```

to-be
```
"dependencies": {
    "@every-analytics/react-analytics-provider": "0.2.0",
```

으로 고치면 되는거 아닌가요? 제가 잘못 생각했나

**Reply by jhlee910609** [reply]
@milooy 앗 제가 이해를 잘못하고 있었던 것 같네요. 3번은 이슈 만들어서 진행볼게요.

**Reply by jhlee910609** [reply]
> 아근데 demo에서 local dependency가 떼지면
> 라이브러리 개발이 불편해지겠네요
> test case 작성 외에 어떻게 live하게 개발할 수 있을까요 🤔

@milooy 이 부분은 레퍼런스가 있는지 좀 찾아볼게요.

**Reply by jhlee910609** [reply]
@milooy https://yarnpkg.com/lang/en/docs/cli/link/ 도움이 좀 될까요? 👀

**Reply by milooy** [reply]
오... 츄라이 해봐요
(yarn2껄로 다시 가져옴 https://yarnpkg.com/cli/link/#gatsby-focus-wrapper)

**Reply by jhlee910609** [reply]
@milooy 도전해봅니다. 헤딩을 많이하게 될 경우 워크로그를 자세히 남긴 후, 지속 공유 드릴게요.
